### PR TITLE
Adding a config block to create views

### PR DESCRIPTION
**Pull request by jedi22**
Sometimes views can be useful to simplify queries that you use a lot. This lets you define views in the config file to be available for queries later.

### CODE REVIEW

**Comment by osqueryer** [comment]
:no_good: The commit acf6ba027c0da875890be000fd54a1ed2d82f831 (Job results: [1585](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1585/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit d84a854d216ebd48140c04e0ba5dcd4474ea1729 (Job results: [1587](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1587/)) failed the code audit or documentation test.

**Comment by theopolis** [comment]
Right, we definitely need something like this.

1. What happens if you change the configuration and remove a view, is the expectation that the view will be removed too?
2. This should include an `osquery_views` table to query the current set of created views and the meta-data about each. Perhaps there are some statistics that can accompany each from SQLite? (just a WAG here)
3. We'll need some documentation to accompany.

1, 3 I'd say are blockers to merging. 2 is a nice-to-have that could be a follow up or wish-list item.

**Comment by jedi22** [comment]
Code Audit seems broken.

**Comment by osqueryer** [comment]
:no_good: The commit 4a3f50afd44400bd7085f73f9542e14aad6c2c47 (Job results: [1588](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1588/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 79098e99c09151d2f4b73a96eede12d13a406447 (Job results: [910](https://jenkins.osquery.io/job/osqueryPullRequestOSX/910/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:no_good: The commit 79098e99c09151d2f4b73a96eede12d13a406447 (Job results: [1589](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1589/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 79098e99c09151d2f4b73a96eede12d13a406447 (Job results: [4770](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4770/)) failed one or more tests (**Linux**).

File: docs/wiki/deployment/configuration.md [hunk 1]
```diff
@@ -411,6 +411,24 @@ Example:
...
+### Views
+
+Views are saved queries expressed as tables. Large subqueries or complex joining logic can often be moved into views allowing you to make your queries more concise.
+
+
+Example:
+```json
+{
+  "views": {
+    "kernel_hashses" : "kernel_hashes" : "select hash.path as kernel_binary, version, hash.sha256 as sha256, hash.sha1 as sha1, hash.md5 as md5 from (select path || '/Contents/MacOS/' as directory, name, version from kernel_extensions) join hash using (directory)"
```
Comment (poppyseedplehzr) [inline docs/wiki/deployment/configuration.md]: Is this line meant to have `kernel_hashes` twice?

Comment (jedi22) [inline docs/wiki/deployment/configuration.md]: Oops.

**Comment by osqueryer** [comment]
:no_good: The commit 7e6eea0f4beb9f35cdf7cf283ac6870ddf7cf95e (Job results: [1601](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1601/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit fbb063ba19f05dd3417b873a079b3a1dd15bce16 (Job results: [1602](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1602/)) failed the code audit or documentation test.

File: docs/wiki/deployment/configuration.md [hunk 2]
```diff
@@ -411,6 +411,24 @@ Example:
 }
 ```
 
+### Views
+
+Views are saved queries expressed as tables. Large subqueries or complex joining logic can often be moved into views allowing you to make your queries more concise.
+
+
```
Comment (theopolis) [inline docs/wiki/deployment/configuration.md]: nit, no triple new line needed

File: osquery/config/parsers/views.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  const auto& views = data_.get_child("views");
+  std::vector<std::string> created_views;
+  for (const auto& view : views) {
+    std::string query = views.get<std::string>(view.first, "");
+    if (query.empty()) {
+      continue;
+    }
+    QueryData results;
+    auto s =
+        osquery::query("CREATE VIEW " + view.first + " AS " + query, results);
```
Comment (theopolis) [inline osquery/config/parsers/views.cpp]: Does this drop and create the view everytime the config is updated? Perhaps this is a broader capability that we need, but I'd assume we'd only "tear down" and recreate views that changed from update-to-update. In the case of the TLS-based config updates, each request (every hour) would be emitting these log lines, after having dropped and created the view.

**Comment by osqueryer** [comment]
:no_good: The commit d251adefa66ad4e38b95ec1898ee63aefd734636 (Job results: [1609](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1609/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 078e2a9925b5bf38a9d9d768b99ea117a1c7fa84 (Job results: [1611](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1611/)) failed the code audit or documentation test.

**Comment by theopolis** [comment]
ok to test

File: osquery/config/parsers/views.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  {
+    std::vector<std::string> old_views_vec;
+    scanDatabaseKeys(kQueries, old_views_vec, kConfigViews);
+    for (const auto& view : old_views_vec) {
+      erase_views.insert(view.substr(kConfigViews.size()));
+    }
+  }
+  QueryData r;
+  for (const auto& view : views) {
+    std::string name = view.first;
```
Comment (theopolis) [inline osquery/config/parsers/views.cpp]: `const auto&`

File: osquery/config/parsers/views.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    // View has been updated
+    osquery::query("DROP VIEW " + name, r);
+    auto s = osquery::query("CREATE VIEW " + name + " AS " + query, r);
+    if (s.ok()) {
+      setDatabaseValue(kQueries, kConfigViews + name, query);
+    } else {
+      LOG(INFO) << "Error creating view (" << name << "): " << s.getMessage();
+    }
+  }
+  for (const auto& old_view : erase_views) {
```
Comment (theopolis) [inline osquery/config/parsers/views.cpp]: Add comment.

File: osquery/config/parsers/views.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                                       const ParserConfig& config) {
+  if (config.count("views") > 0) {
+    data_ = pt::ptree();
+    data_.put_child("views", config.at("views"));
+  }
+  const auto& views = data_.get_child("views");
+  std::vector<std::string> created_views;
+  std::set<std::string> erase_views;
+  {
+    std::vector<std::string> old_views_vec;
```
Comment (theopolis) [inline osquery/config/parsers/views.cpp]: Add comment for data structure change

File: osquery/tests/test_util.cpp [hunk 7]
```diff
@@ -103,9 +103,9 @@ void shutdownTesting() {
   Initializer::platformTeardown();
 }
 
-std::map<std::string, std::string> getTestConfigMap() {
+std::map<std::string, std::string> getTestConfigMap(std::string file) {
```
Comment (theopolis) [inline osquery/tests/test_util.cpp]: `const std::string& file`

File: osquery/tests/test_util.h [hunk 8]
```diff
@@ -76,7 +76,8 @@ extern const char* kExpectedExtensionArgs[];
 extern const size_t kExpectedExtensionArgsCount;
 
 // Get an example generate config with one static source name to JSON content.
-std::map<std::string, std::string> getTestConfigMap();
+std::map<std::string, std::string> getTestConfigMap(
+    std::string file = "test_parse_items.conf");
```
Comment (theopolis) [inline osquery/tests/test_util.h]: `const std::string&`

File: osquery/config/parsers/views.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <set>
```
Comment (theopolis) [inline osquery/config/parsers/views.cpp]: nit newline

File: docs/wiki/deployment/configuration.md [hunk 10]
```diff
@@ -411,6 +411,23 @@ Example:
...
 
+### Views
+
+Views are saved queries expressed as tables. Large subqueries or complex joining logic can often be moved into views allowing you to make your queries more concise.
+
+Example:
+```json
+{
+  "views": {
+    "kernel_hashses" : "select hash.path as kernel_binary, version, hash.sha256 as sha256, hash.sha1 as sha1, hash.md5 as md5 from (select path || '/Contents/MacOS/' as directory, name, version from kernel_extensions) join hash using (directory)"
```
Comment (reedloden) [inline docs/wiki/deployment/configuration.md]: `kernel_hashes`

**Pull request merged**